storagenode/gracefulexit/worker.go

```diff
@@ -4,48 +4,248 @@
+		}
+
+		switch msg := response.GetMessage().(type) {
+		case *pb.SatelliteMessage_NotReady:
+			break // wait until next worker execution
+		case *pb.SatelliteMessage_TransferPiece:
+			pieceID := msg.TransferPiece.OriginalPieceId
+			reader, err := worker.store.Reader(ctx, satelliteID, pieceID)
+			if err != nil {
+				transferErr := pb.TransferFailed_UNKNOWN
+				if errs.Is(err, os.ErrNotExist) {
+					transferErr = pb.TransferFailed_NOT_FOUND
+				}
+				worker.log.Error("failed to get piece reader.", zap.Stringer("satellite ID", satelliteID), zap.Stringer("piece ID", pieceID), zap.Error(errs.Wrap(err)))
+				worker.handleFailure(ctx, transferErr, pieceID, satelliteID, c.Send)
+				continue
+			}
+
+			addrLimit := msg.TransferPiece.GetAddressedOrderLimit()
+			pk := msg.TransferPiece.PrivateKey
+
+			originalHash, originalOrderLimit, err := worker.getHashAndLimit(ctx, reader, addrLimit.GetLimit())
+			if err != nil {
+				worker.log.Error("failed to get piece hash and order limit.", zap.Stringer("satellite ID", satelliteID), zap.Stringer("piece ID", pieceID), zap.Error(errs.Wrap(err)))
+				worker.handleFailure(ctx, pb.TransferFailed_UNKNOWN, pieceID, satelliteID, c.Send)
+				continue
+			}
+
+			putCtx, cancel := context.WithCancel(ctx)
+			defer cancel()
+
+			// TODO what's the typical expiration setting?
+			pieceHash, err := worker.ecclient.PutPiece(putCtx, ctx, addrLimit, pk, reader, time.Now().Add(time.Second*600))
+			if err != nil {
+				worker.log.Error("failed to put piece.", zap.Stringer("satellite ID", satelliteID), zap.Stringer("piece ID", pieceID), zap.Error(errs.Wrap(err)))
+				// TODO look at error type to decide on the transfer error
+				worker.handleFailure(ctx, pb.TransferFailed_STORAGE_NODE_UNAVAILABLE, pieceID, satelliteID, c.Send)
+
+				continue
+			}
+
+			err = verifyPieceHash(ctx, addrLimit.GetLimit(), pieceHash, originalHash.Hash)
+			if err != nil {
+				worker.log.Error("failed hash verification.", zap.Stringer("satellite ID", satelliteID), zap.Stringer("piece ID", pieceID), zap.Error(errs.Wrap(err)))
+				worker.handleFailure(ctx, pb.TransferFailed_HASH_VERIFICATION, pieceID, satelliteID, c.Send)
+
+				continue
+			}
+
+			success := &pb.StorageNodeMessage{
+				Message: &pb.StorageNodeMessage_Succeeded{
+					Succeeded: &pb.TransferSucceeded{
+						OriginalPieceId:      msg.TransferPiece.OriginalPieceId,
+						OriginalPieceHash:    &pb.PieceHash{PieceId: msg.TransferPiece.OriginalPieceId},
+						OriginalOrderLimit:   &originalOrderLimit,
+						ReplacementPieceHash: pieceHash,
+					},
+				},
+			}
+			err = c.Send(success)
+			if err != nil {
+				return errs.Wrap(err)
+			}
+		case *pb.SatelliteMessage_DeletePiece:
+			pieceID := msg.DeletePiece.OriginalPieceId
+			err := worker.store.Delete(ctx, satelliteID, pieceID)
+			if err != nil {
+				worker.log.Error("failed to delete piece.", zap.Stringer("satellite ID", satelliteID), zap.Stringer("piece ID", pieceID), zap.Error(errs.Wrap(err)))
```

Comment:
Should we break here to wait for the next worker execution?

Reply:
I was thinking we'd just move on to the next message, but maybe I'm missing something. In addition, the storage node should purge any remaining pieces at the end of the overall graceful exit process anyway.